test/mcmc/ess.jl

```diff
@@ -43,41 +45,49 @@ using Turing
     end
 
     @testset "ESS inference" begin
-        Random.seed!(1)
-        chain = sample(demo_default, ESS(), 5_000)
-        check_numerical(chain, [:m], [0.8]; atol=0.1)
-
-        Random.seed!(1)
-        chain = sample(demodot_default, ESS(), 5_000)
-        check_numerical(chain, ["m[1]", "m[2]"], [0.0, 0.8]; atol=0.1)
-
-        Random.seed!(100)
-        alg = Gibbs(CSMC(15, :s), ESS(:m))
-        chain = sample(gdemo(1.5, 2.0), alg, 10_000)
-        check_numerical(chain, [:s, :m], [49 / 24, 7 / 6]; atol=0.1)
-
-        # MoGtest
-        Random.seed!(125)
-        alg = Gibbs(CSMC(15, :z1, :z2, :z3, :z4), ESS(:mu1), ESS(:mu2))
-        chain = sample(MoGtest_default, alg, 6000)
-        check_MoGtest_default(chain; atol=0.1)
-
-        # Different "equivalent" models.
-        # NOTE: Because `ESS` only supports "single" variables with
-        # Gaussian priors, we restrict ourselves to this subspace by conditioning
-        # on the non-Gaussian variables in `DEMO_MODELS`.
-        models_conditioned = map(DynamicPPL.TestUtils.DEMO_MODELS) do model
-            # Condition on the non-Gaussian random variables.
-            model | (s=DynamicPPL.TestUtils.posterior_mean(model).s,)
+        @info "Starting ESS inference tests"
+        rng = StableRNG(23)
+
+        @testset "demo_default" begin
+            chain = sample(copy(rng), demo_default, ESS(), 5_000)
```

Comment:
Is there a reason you prefer using `copy(rng)` over instantiating a fresh one with `StableRNG(23)`?

I feel like I'd prefer the latter, because even though it's more verbose it keeps the state self-contained to a testset. Otherwise for example someone could accidentally mutate the rng between testsets and then subsequent tests would go wrong.

Reply:
The only reasons were the verbosity, and ability to change the seed globally if desired. I do see your point though.

Would you prefer having `sample(StableRNG(23)` or `sample(StableRNG(seed)`, where `seed` would be defined at the top? I think the latter could be a best-of-both-worlds.

Reply:
Hmm... in a purist sense `StableRNG(23)` appeals to me because the 23 in one testset isn't the same as the 23 in the other testset. The only case where the seed must be the same is if you're trying to prove that sampling is deterministic i.e. `seed = 23; sample(StableRNG(seed), ...) == sample(StableRNG(seed), ...)`.

Sprinkling that over the entire test suite would be ugly, though, so happy to go with the latter, and as you say that preserves our ability to change everything globally in the way we could with `Random.seed!`.

Also, is there a particular reason for 23? 😄

Reply:
I agree that there's no need for the seeds to be exactly the same, but it's rather being able to have this cascade of changing a global seed and having everything be affected. Handy for instance if we want to test how brittle our whole test suite is to changes in randomisation.

> Also, is there a particular reason for 23? 😄

It is known to be the most synchronistic number, which means that the pseudorandom numbers generated will have more invisible correlations in them, making the tests more robust.

Reply:
🤯 And here I am just choosing my favourite Pokemon's dex number

Reply:
Now doing `StableRNG(seed)` all over.